architecture/2025-04-compute-extensibility.md

```diff
@@ -0,0 +1,474 @@
+
+The primary criterion for deciding between option 1 and 2 is whether we want the application model to be represented as Radius-provided UDTs, or if we want to continue to have the core resource types that represent the application model to be built-in.
+
+Both options have technical risk, primarily in the implementation of Kubernetes provisioning with recipes and the associated separation of provisioning code from the Radius application. Option 2 has lower initial risk and faster time to initial release because the scope of the initial release is smaller.
+
+## Terms and definitions
+
+| Term | Definition |
+|------|------------|
+| **Compute platform** | An environment where applications can be deployed and run, such as Kubernetes, Azure Container Instances (ACI), etc. |
+| **Core types** | Built-in resource types provided by Radius, including `containers`, `gateways`, `secretStores`, `environments`, and `applications`. |
+| **User-defined type (UDT)** | A custom resource type defined separately from Radius core types and loaded into Radius without requiring a new Radius release. |
+| **Recipe** | A set of instructions that provisions a Radius resource to a Radius environment. Recipes are implemented in Bicep or Terraform. |
+| **Resource Provider (RP)** | A component responsible for handling create, read, update, delete, and list (CRUDL) operations for a specific resource type. |
+
+### Goals
+
+* Provide platform engineers with the ability to deploy to specific platforms other than Radius, like ACI.
+* Provide a recipe-based platform engineering experience that is consistent for user-defined types and core types.
+* Expand the ability to provision a single application definition to multiple clouds by adding the capability to provision to multiple compute platforms like ACI and other serverless technologies.
+
+### Non-goals
+
+* Running the Radius control plane on a non-Kubernetes platform
+* Changes to portable types
+
+## Principles
+
+Radius extensibility should enable creating extensions that are:
+
+| Principle | Description |
+|-----------|-------------|
+| **Independently upgradeable in a runtime environment** | Recipes are independently upgradeable. |
+| **Isolated and over a network protocol** | Recipes are currently implemented through Bicep and Terraform. Both run locally and communicate to the target platform over secure network protocols. |
+| **Strongly typed and support versioning** | Recipes support versioning through the use of OCI-compliant registries. They are not strongly typed in the sense of having compile-time validation, but they can be validated upon registration in an environment as having the correct input parameters and output properties. |
+
+### User scenarios
+
+#### Configure a non-Kubernetes Radius environment
+
+As a platform engineer I can initialize a new Radius environment that is connected to a non-Kubernetes compute platform so that Radius can authenticate and deploy to non-Kubernetes platforms.
```

Comment:
We should think about how existing users can migrate to this new version of Radius. We may need to introduce a `rad migrate` command.

Reply:
Good point. Hopefully migrating is simple because the UDT versions of the core resource types would be similar to the existing core types.